spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/AbstractStreamService.java

```diff
@@ -0,0 +1,185 @@
+	}
+
+	public void updateStream(String streamName, String releaseName, PackageIdentifier packageIdenfier,
+			Map<String, String> updateProperties) {
+		StreamDeployment streamDeployment = this.streamDeploymentRepository.findOne(streamName);
+		if (streamDeployment == null) {
+			throw new NoSuchStreamDeploymentException(streamName);
+		}
+		if (this.streamDeployer != StreamDeployers.valueOf(streamDeployment.getDeployerName())) {
+			throw new IncompatibleStreamDeployerException(streamDeployer.name());
+		}
+
+		doUpdateStream(streamName, releaseName, packageIdenfier, updateProperties);
+	}
+
+	protected abstract void doUpdateStream(String streamName, String releaseName, PackageIdentifier packageIdenfier,
+			Map<String, String> updateProperties);
+
+	@Override
+	public void rollbackStream(String streamName, int releaseVersion) {
+		Assert.isTrue(StringUtils.hasText(streamName), "Stream name must not be null");
+		StreamDeployment streamDeployment = this.streamDeploymentRepository.findOne(streamName);
+		if (streamDeployment == null) {
+			throw new NoSuchStreamDeploymentException(streamName);
+		}
+		if (this.streamDeployer != StreamDeployers.valueOf(streamDeployment.getDeployerName())) {
+			throw new IncompatibleStreamDeployerException(streamDeployer.name());
+		}
+		doRollbackStream(streamName, releaseVersion);
+	}
+
+	protected abstract void doRollbackStream(String streamName, int releaseVersion);
+
+	protected StreamDefinition createStreamDefinitionForDeploy(String name) {
+		StreamDefinition streamDefinition = this.streamDefinitionRepository.findOne(name);
+		if (streamDefinition == null) {
+			throw new NoSuchStreamDefinitionException(name);
+		}
+
+		String status = this.doCalculateStreamState(name);
+
+		if (DeploymentState.deployed.equals(DeploymentState.valueOf(status))) {
+			throw new StreamAlreadyDeployedException(name);
+		}
+		else if (DeploymentState.deploying.equals(DeploymentState.valueOf(status))) {
+			throw new StreamAlreadyDeployingException(name);
+		}
+		return streamDefinition;
+	}
+
+	protected abstract String doCalculateStreamState(String name);
+
+	// State
+	@Override
+	public Map<StreamDefinition, DeploymentState> state(List<StreamDefinition> streamDefinitions) {
+		Map<StreamDefinition, DeploymentState> states = new HashMap<>();
+		List<StreamDefinition> deployerSpecificStreamDefinitions = new ArrayList<>();
+		for (StreamDefinition streamDefinition : streamDefinitions) {
+			StreamDeployment streamDeployment = this.streamDeploymentRepository.findOne(streamDefinition.getName());
+			if (streamDeployment == null) {
+				states.put(streamDefinition, DeploymentState.unknown);
+			}
+			else if (this.streamDeployer == StreamDeployers.valueOf(streamDeployment.getDeployerName())) {
+				deployerSpecificStreamDefinitions.add(streamDefinition);
+			}
+			else {
+				logger.warn("Deployer incompatible stream definition:" + streamDefinition);
```

Comment:
I think we need to throw an exception here as it is not expected to have an entry of deployer other than `stream` and `skipper`.

Reply:
right